scipy/sparse/linalg/interface.py

```diff
@@ -68,43 +127,63 @@ class LinearOperator(object):
     array([ 2.,  3.])
 
     """
-    def __init__(self, shape, matvec, rmatvec=None, matmat=None, dtype=None):
+    def __new__(cls, *args, **kwargs):
+        if cls is LinearOperator:
+            # Operate as _CustomLinearOperator factory.
+            return _CustomLinearOperator(*args, **kwargs)
+        else:
+            obj = super(LinearOperator, cls).__new__(cls)
 
-        shape = tuple(shape)
+            if (type(obj)._matvec == LinearOperator._matvec
+                    and type(obj)._matmat == LinearOperator._matmat):
+                raise TypeError("LinearOperator subclass should implement"
+                                " at least one of _matvec and _matmat.")
 
-        if not isshape(shape):
-            raise ValueError('invalid shape')
+            obj.__init__(*args, **kwargs)
+            return obj
 
-        self.shape = shape
-        self._matvec = matvec
-        self.args = ()
+    def __init__(self, dtype, shape):
+        """Initialize this LinearOperator.
 
-        if rmatvec is None:
-            def rmatvec(v):
-                raise NotImplementedError('rmatvec is not defined')
-            self.rmatvec = rmatvec
-        else:
-            self.rmatvec = rmatvec
+        To be called by subclasses. ``dtype`` may be None; ``shape`` should
+        be convertible to a length-2 tuple.
+        """
+        if dtype is not None:
+            dtype = np.dtype(dtype)
 
-        if matmat is not None:
-            # matvec each column of V
-            self._matmat = matmat
+        shape = tuple(shape)
+        if not isshape(shape):
+            raise ValueError("invalid shape %r (must be 2-d)" % shape)
 
-        if dtype is not None:
-            self.dtype = np.dtype(dtype)
+        self.dtype = dtype
```

Comment:
I've attempted to make the changes suggested above in https://github.com/scipy/scipy/pull/4316 but it seems to not work.